rasa/utils/tensorflow/models.py

```diff
@@ -260,14 +266,37 @@ def rasa_predict(self, model_data: RasaModelData) -> Dict[Text, tf.Tensor]:
         batch_in = RasaBatchDataGenerator.prepare_batch(model_data.data)
 
         if self._run_eagerly:
-            return tf_utils.to_numpy_or_python_type(self.predict_step(batch_in))
+            outputs = tf_utils.to_numpy_or_python_type(self.predict_step(batch_in))
+            outputs[DIAGNOSTIC_DATA] = self._empty_lists_to_none_in_dict(
+                outputs[DIAGNOSTIC_DATA]
+            )
+            return outputs
 
         if self._tf_predict_step is None:
             self._tf_predict_step = tf.function(
                 self.predict_step, input_signature=self._dynamic_signature(batch_in)
             )
 
-        return tf_utils.to_numpy_or_python_type(self._tf_predict_step(batch_in))
+        outputs = tf_utils.to_numpy_or_python_type(self._tf_predict_step(batch_in))
+        outputs[DIAGNOSTIC_DATA] = self._empty_lists_to_none_in_dict(
+            outputs[DIAGNOSTIC_DATA]
+        )
+        return outputs
+
+    @staticmethod
+    def _empty_lists_to_none_in_dict(input_dict: Dict[Text, Any]) -> Dict[Text, Any]:
```

Comment:
_recurse needs type annotations and I think reads much simpler when formulated this way:

```
    @staticmethod
    def _empty_lists_to_none_in_dict(input_dict: Dict[Text, Any]) -> Dict[Text, Any]:
        """Recursively replaces empty list or np array with None in a dictionary."""

        def _recurse(x):
            if isinstance(x, dict):
                return {k: _recurse(v) for k, v in x.items()}
            elif (isinstance(x, list) or isinstance(x, np.ndarray)) and np.size(x) == 0:
                return None
            else:
                return x

        return _recurse(input_dict)
```

Reply:
Agree, I'll make these changes but I'll keep the last `return` in `_recurse` outside the if-else.